File: quicksort Multithreaded/QuickSort.java
import java.util.*;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveAction;

public class QuickSort {

    private static class QuickSortTask extends RecursiveAction {
        private final int[] arr;
        private final int left;
        private final int right;

        public QuickSortTask(int[] arr, int left, int right) {
            this.arr = arr;
            this.left = left;
            this.right = right;
        }

        @Override
        protected void compute() {
            if (left >= right) {
                return;
            }

            int pivot = partition(arr, left, right);

            QuickSortTask leftTask = new QuickSortTask(arr, left, pivot - 1);
            QuickSortTask rightTask = new QuickSortTask(arr, pivot + 1, right);

            invokeAll(leftTask, rightTask);
        }

        private int partition(int[] arr, int left, int right) {
            int pivot = arr[right];
            int i = left - 1;

            for (int j = left; j < right; j++) {
                if (arr[j] < pivot) {
                    i++;
                    swap(arr, i, j);
                }
            }

            swap(arr, i + 1, right);
            return i + 1;
        }

        private void swap(int[] arr, int i, int j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void main(String[] args) {
        long startTime;
        long endTime;


        int[] largeArray1 = new int[7000000];
        
        for(int i = 7000000; i<0; i--){
            largeArray1[7000000-i] = i;
        }

        startTime = System.currentTimeMillis();

        ForkJoinPool pool = new ForkJoinPool();
        pool.invoke(new QuickSortTask(arr, 0, arr.length - 1));
        endTime = System.currentTimeMillis();
        System.out.println("Sorted array: "+(endTime-startTime));
    }
}

File: quicksort Multithreaded/QuickSort.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <random>

using namespace std;
void quicksort(std::vector<int>& arr, int left, int right) {
    if (left >= right) {
        return;
    }

    int pivot = arr[right];
    int i = left - 1;

    for (int j = left; j < right; j++) {
        if (arr[j] < pivot) {
            i++;
            std::swap(arr[i], arr[j]);
        }
    }

    std::swap(arr[i + 1], arr[right]);
    pivot = i + 1;

    std::thread left_thread(quicksort, std::ref(arr), left, pivot - 1);
    std::thread right_thread(quicksort, std::ref(arr), pivot + 1, right);

    left_thread.join();
    right_thread.join();
}

int main() {
   const int size = 10;
    const int range_min = 0;
    const int range_max = 100;

   random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(range_min, range_max);

    std::vector<int> arr(size);
    for (int i = 0; i < size; i++) {
        arr[i] = dis(gen);
    }
    for (int i = 0; i < size; i++) {
        cout << arr[i] << " ";
    }
    cout<<endl;


    quicksort(arr, 0, size - 1);
    for (int i = 0; i < size; i++) {
        cout << arr[i] << " ";
    }

    return 0;
}
